File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/map/MapViewOnTouchListener.java
package com.valentingrigorean.arcgis_maps_flutter.map;

import android.content.Context;
import android.util.Log;
import android.view.MotionEvent;

import com.esri.arcgisruntime.concurrent.ListenableFuture;
import com.esri.arcgisruntime.geometry.Point;
import com.esri.arcgisruntime.mapping.view.DefaultMapViewOnTouchListener;
import com.esri.arcgisruntime.mapping.view.Graphic;
import com.esri.arcgisruntime.mapping.view.IdentifyGraphicsOverlayResult;
import com.esri.arcgisruntime.mapping.view.IdentifyLayerResult;
import com.esri.arcgisruntime.mapping.view.MapView;
import com.valentingrigorean.arcgis_maps_flutter.Convert;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import io.flutter.plugin.common.MethodChannel;

public class MapViewOnTouchListener extends DefaultMapViewOnTouchListener {

    private static final String TAG = "MapViewOnTouchListener";

    private final ArrayList<MapTouchGraphicDelegate> mapTouchGraphicDelegates = new ArrayList<>();
    private final MethodChannel methodChannel;

    private ListenableFuture<List<IdentifyGraphicsOverlayResult>> graphicHandler;
    private ListenableFuture<List<IdentifyLayerResult>> layerHandler;
    private boolean trackIdentityLayers;

    private final FlutterMapViewDelegate flutterMapViewDelegate;

    public MapViewOnTouchListener(Context context, FlutterMapViewDelegate flutterMapViewDelegate, MethodChannel methodChannel) {
        super(context, flutterMapViewDelegate.getMapView());
        this.methodChannel = methodChannel;
        this.flutterMapViewDelegate = flutterMapViewDelegate;
    }

    public void setTrackIdentityLayers(boolean trackIdentityLayers) {
        this.trackIdentityLayers = trackIdentityLayers;
    }

    public void addGraphicDelegate(MapTouchGraphicDelegate mapTouchGraphicDelegate) {
        if (!mapTouchGraphicDelegates.contains(mapTouchGraphicDelegate))
            mapTouchGraphicDelegates.add(mapTouchGraphicDelegate);
    }

    public void clearAllDelegates() {
        mapTouchGraphicDelegates.clear();
    }

    @Override
    public boolean onSingleTapConfirmed(MotionEvent e) {

        clearHandlers();

        final android.graphics.Point screenPoint = new android.graphics.Point((int) e.getX(), (int) e.getY());

        if (canConsumeGraphics()) {
            Log.d(TAG, "onSingleTapConfirmed: identifyGraphicsOverlays");
            identifyGraphicsOverlays(screenPoint);
        } else if (trackIdentityLayers) {
            Log.d(TAG, "onSingleTapConfirmed: identifyLayers");
            identifyLayers(screenPoint);
        } else {
            sendOnMapTap(screenPoint);
        }

        return true;
    }

    @Override
    public void onLongPress(MotionEvent e) {
        super.onLongPress(e);
        final android.graphics.Point screenPoint = new android.graphics.Point((int) e.getX(), (int) e.getY());
        sendOnMapLongPress(screenPoint);
    }

    private void identifyGraphicsOverlays(android.graphics.Point screenPoint) {
        graphicHandler = mMapView.identifyGraphicsOverlaysAsync(screenPoint, 12, false);
        graphicHandler.addDoneListener(() -> {
            try {
                final List<IdentifyGraphicsOverlayResult> results = graphicHandler.get();
                graphicHandler = null;
                Log.d(TAG, "identifyGraphicsOverlays: " + results.size() + " found.");
                if (onTapCompleted(results, screenPoint)) {
                    return;
                }
                if (trackIdentityLayers) {
                    identifyLayers(screenPoint);
                } else {
                    sendOnMapTap(screenPoint);
                }
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        });
    }

    private void identifyLayers(android.graphics.Point screenPoint) {
        layerHandler = mMapView.identifyLayersAsync(screenPoint, 10, false);
        layerHandler.addDoneListener(() -> {
            try {
                final List<IdentifyLayerResult> results = layerHandler.get();
                layerHandler = null;
                Log.d(TAG, "identifyLayers: " + results.size() + " found.");
                if (results.size() == 0) {
                    sendOnMapTap(screenPoint);
                    return;
                }
                final Object jsonResults = Convert.identifyLayerResultsToJson(results);
                final Point position = mMapView.screenToLocation(screenPoint);
                final HashMap<String, Object> data = new HashMap<>(3);
                data.put("results", jsonResults);
                data.put("screenPoint", Convert.pointToJson(screenPoint));
                data.put("position", Convert.geometryToJson(position));

                methodChannel.invokeMethod("map#onIdentifyLayers", data);
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        });
    }

    private boolean onTapCompleted(List<IdentifyGraphicsOverlayResult> results, android.graphics.Point screenPoint) {
        if (results != null) {
            for (final IdentifyGraphicsOverlayResult result : results) {
                for (final MapTouchGraphicDelegate touchDelegate : mapTouchGraphicDelegates) {
                    for (final Graphic graphic : result.getGraphics()) {
                        if (touchDelegate.didHandleGraphic(graphic)) {
                            return true;
                        }
                    }
                }
            }
        }
        return false;
    }

    private void sendOnMapTap(android.graphics.Point screenPoint) {
        final Point mapPoint = mMapView.screenToLocation(screenPoint);
        if (mapPoint == null) {
            return;
        }
        final HashMap<String, Object> data = new HashMap<>(1);
        data.put("position", Convert.geometryToJson(mapPoint));
        data.put("screenPoint", Convert.pointToJson(screenPoint));
        methodChannel.invokeMethod("map#onTap", data);
    }

    private void sendOnMapLongPress(android.graphics.Point screenPoint) {
        final Point mapPoint = mMapView.screenToLocation(screenPoint);
        if (mapPoint == null) {
            return;
        }

        final HashMap<String, Object> data = new HashMap<>(1);
        data.put("position", Convert.geometryToJson(mapPoint));
        data.put("screenPoint", Convert.pointToJson(screenPoint));
        methodChannel.invokeMethod("map#onLongPress", data);
    }

    private void clearHandlers() {
        if (graphicHandler != null) {
            graphicHandler.cancel(true);
            graphicHandler = null;
        }
        if (layerHandler != null) {
            layerHandler.cancel(true);
            layerHandler = null;
        }
    }

    private boolean canConsumeGraphics() {
        for (final MapTouchGraphicDelegate touchDelegate : mapTouchGraphicDelegates) {
            if (touchDelegate.canConsumeTaps()) {
                return true;
            }
        }
        return false;
    }
}

File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/map/MapViewOnTouchListener.kt
package com.valentingrigorean.arcgis_maps_flutter.map

import android.util.Log
import com.arcgismaps.geometry.Point
import com.arcgismaps.mapping.view.IdentifyGraphicsOverlayResult
import com.arcgismaps.mapping.view.MapView
import com.arcgismaps.mapping.view.ScreenCoordinate
import com.valentingrigorean.arcgis_maps_flutter.convert.geometry.toFlutterJson
import com.valentingrigorean.arcgis_maps_flutter.convert.mapping.view.toFlutterJson
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch

private enum class MessageType {
    OnTap,
    OnLongPress,
    OnLongPressEnd,
}

class MapViewOnTouchListener(
    private val mapView: MapView,
    private val methodChannel: MethodChannel,
    private val scope: CoroutineScope,
) {
    private val mapTouchGraphicDelegates = ArrayList<MapTouchGraphicDelegate>()
    private var graphicJob: Job? = null
    private var layerJob: Job? = null
    private var isLongPress = false
    var trackIdentityLayers = false


    init {
        mapView.onSingleTapConfirmed.onEach {
            clearHandlers()
            if (mapView.map == null) {
                return@onEach
            }

            if (canConsumeGraphics()) {
                Log.d(TAG, "onSingleTapConfirmed: identifyGraphicsOverlays")
                identifyGraphicsOverlays(it.screenCoordinate, it.mapPoint)
            } else if (trackIdentityLayers) {
                Log.d(TAG, "onSingleTapConfirmed: identifyLayers")
                identifyLayers(it.screenCoordinate, it.mapPoint)
            } else {
                sendMessage(it.screenCoordinate, it.mapPoint, MessageType.OnTap)
            }
        }.launchIn(scope)

        mapView.onLongPress.onEach {
            isLongPress = true
            sendMessage(it.screenCoordinate, it.mapPoint, MessageType.OnLongPress)
        }.launchIn(scope)

        mapView.onUp.onEach {
            if (isLongPress) {
                isLongPress = false
                sendMessage(it.screenCoordinate, it.mapPoint, MessageType.OnLongPressEnd)
            }
        }.launchIn(scope)
    }

    fun addGraphicDelegate(mapTouchGraphicDelegate: MapTouchGraphicDelegate) {
        if (!mapTouchGraphicDelegates.contains(mapTouchGraphicDelegate)) mapTouchGraphicDelegates.add(
            mapTouchGraphicDelegate
        )
    }

    fun clearAllDelegates() {
        mapTouchGraphicDelegates.clear()
    }


    private fun identifyGraphicsOverlays(screenPoint: ScreenCoordinate, point: Point?) {
        graphicJob = scope.launch {
            mapView.identifyGraphicsOverlays(screenPoint, 10.0, false, 10).onSuccess {
                if (onTapCompleted(it)) {
                    return@launch
                }
                if (trackIdentityLayers) {
                    identifyLayers(screenPoint, point)
                } else {
                    sendMessage(screenPoint, point, MessageType.OnTap)
                }
            }
        }
    }

    private fun identifyLayers(screenPoint: ScreenCoordinate, point: Point?) {
        layerJob = scope.launch {
            mapView.identifyLayers(screenPoint, 10.0, false, 10).onSuccess { results ->
                if (results.isEmpty()) {
                    sendMessage(screenPoint, point, MessageType.OnTap)
                    return@launch
                }

                val data = mapOf(
                    "screenPoint" to arrayListOf(screenPoint.x, screenPoint.y),
                    "position" to point?.toFlutterJson(),
                    "results" to results.map { it.toFlutterJson() }
                )
                methodChannel.invokeMethod("map#onIdentifyLayers", data);
            }
        }
    }

    private fun onTapCompleted(
        results: List<IdentifyGraphicsOverlayResult>,
    ): Boolean {
        for (result in results) {
            for (touchDelegate in mapTouchGraphicDelegates) {
                for (graphic in result.graphics) {
                    if (touchDelegate.didHandleGraphic(graphic)) {
                        return true
                    }
                }
            }
        }
        return false
    }

    private fun sendMessage(
        screenPoint: ScreenCoordinate,
        point: Point?,
        messageType: MessageType
    ) {
        val data = HashMap<String, Any?>(2)
        data["position"] = point?.toFlutterJson()
        data["screenPoint"] = arrayListOf(screenPoint.x, screenPoint.y)
        when (messageType) {
            MessageType.OnTap -> methodChannel.invokeMethod("map#onTap", data)
            MessageType.OnLongPress -> methodChannel.invokeMethod("map#onLongPress", data)
            MessageType.OnLongPressEnd -> methodChannel.invokeMethod("map#onLongPressEnd", data)
        }
    }

    private fun clearHandlers() {
        graphicJob?.cancel()
        graphicJob = null
        layerJob?.cancel()
        layerJob = null
    }

    private fun canConsumeGraphics(): Boolean {
        for (touchDelegate in mapTouchGraphicDelegates) {
            if (touchDelegate.canConsumeTaps()) {
                return true
            }
        }
        return false
    }

    companion object {
        private const val TAG = "MapViewOnTouchListener"
    }
}